enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,742 @@
+	// +patchStrategy=merge
+	// +optional
+	ResolvedNames []DNSNameResolverStatusItem `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
+}
+
+// DNSNameResolverStatusItem describes the details of a resolved DNS name.
+type DNSNameResolverStatusItem struct {
+	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	DNSName string `json:"dnsName"`
+	// ips contains the list of IP addresses associated with the dnsName.
+	// +kubebuilder:validation:Required
+	// +listType=set
+	IPs []string `json:"ips"`
+	// ttlSeconds is the minimum time-to-live value among all the IP addresses.
+	// +kubebuilder:validation:Required
+	TTLSeconds int32 `json:"ttlSeconds"`
+	// lastLookupTime is the timestamp when the last DNS lookup was completed.
+	// +kubebuilder:validation:Required
+	LastLookupTime *metav1.Time `json:"lastLookupTime"`
+	// resolutionFailures keeps the count of how many times the DNS resolution failed for the
+	// dnsName field. If the DNS resolution succeeds then the field will be set to zero. Upon
+	// every failure, the value of the field will be incremented by one. Upon reaching a threshold
+	// value, the details about the DNS name will be removed.
+	ResolutionFailures int `json:"resolutionFailures,omitempty"`
+	// conditions provide information about the state of the DNS name.
+	//
+	// These are the supported conditions:
+	//
+	//   * Degraded
+	//   - True if the following conditions are met:
+	//     * The last DNS name resolution failed.
+	//   - False if any of those conditions are unsatisfied.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+````
+
+### Workflow Description
+
+The workflows for the different components related to DNSNameResolver events are explained in this section. The workflow
+for those events is shown in the following diagram with an example `DNSNameResolver` CR:
+
+![Workflow](./coredns-egressfirewall-integration.png)
+
+#### Workflow of OVN-K cluster manager
+
+The following pseudocode explains the workflow of the OVN-K cluster manager:
+```shell
+watch(EgressFirewall)
+
+if match(event.type, "create"):
+
+	for each rule in obj.egress:
+		if rule.to.dnsName != ""  && check_dns_name_resolver_exists(rule.to.dnsName) == false:
+			create_dns_name_resolver(rule.to.dnsName)
+
+else if match(event.type, "delete"):
+
+	for each rule in obj.egress:
+		if rule.to.dnsName != "" && check_dns_name_used_in_any_egress_firewall(rule.to.dnsName) == false:
+			delete_dns_name_resolver(rule.to.dnsName)
+```
+
+* The OVN-K cluster manager will watch for events related to `EgressFirewall` objects. (Note: Currently, on update of an `EgressFirewall` object, it is
+first deleted and then created again by the OVN-K master(s). Thus, only the create and delete events are handled here.)
```

Comment:
I wanted to get some clarifications on this.

- Why is an update event handled by deleting and creating the EgressFirewall object?
- Since cluster manager is also part of ovn-k, wouldn't it be better that the masters and the cluster manager handle the EgressFirewall objects similarly?

From the point of view of cluster manager, if it were to handle the update event, then the cluster manager will have to do the following:

1. Spend CPU cycles to process the EgressFirewall object to get the changes in the DNS names and create/delete DNSNameResolver objects based on the derived changes in the EgressFirewall object
2. Then get a delete event, process the object again and delete DNSNameResolver objects, if required
3. Then get a create event, process the object again and create DNSNameResolver objects, if required

The time and resources spent on processing the update event will not be useful as the same changes will again be brought in by the following delete and create events.

Reply:
Maybe I am explaining this in a confusing way, but there will be no extra delete/create event, so steps 2 and 3 won't happen. Inside ovn-k master there will also only be an update event, that will be handled by deleting and re-creating ACLs instead of evaluating the diff and only applying the real changes. Does it make sense?

Reply:
Ok. Got it. I will update the workflow.

Reply:
Added the details for update event.